cdap-ui/app/cdap/components/Reports/Customizer/TimeRangeSelector/TimeRangePopover.js

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import Popover from 'components/Popover';
+import IconSVG from 'components/IconSVG';
+import TimeRangePicker from 'components/TimeRangePicker';
+import {connect} from 'react-redux';
+import {ReportsActions} from 'components/Reports/store/ReportsStore';
+
+class TimeRangePopoverView extends Component {
+  static propTypes = {
+    onApply: PropTypes.func,
+    start: PropTypes.number,
+    end: PropTypes.number,
+    selection: PropTypes.selection
+  };
+
+  state = {
+    selection: this.props.selection,
+    start: this.props.start,
+    end: this.props.end
+  };
+
+  componentWillReceiveProps(nextProps) {
+    this.setState({
```

Comment:
nit: maybe do `if (nextProps.selection !== this.props.selection || nextProps.start !== this.props.start || nextProps.end !== this.props.end)` to avoid unnecessary re-rendering?

Reply:
it's not needed. The only time this props will get updated is when you hit apply, at which point the popup is closed. The component update is handled inside until you hit apply, then it will go to the store.